### PR TITLE
Initial commit for Retry module

### PR DESCRIPTION
**Pull request by MVenglarcik**

### CODE REVIEW

File: Retry/HELP.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# Getting Started
```
Comment (typekpb) [inline Retry/HELP.md]: convention says, file should be named README.md, to be able to show it directly in the Github UI once opening the folder

File: Retry/src/main/java/org/techsalad/retry/controller/SaladsController.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+  @PostConstruct
+  public void postConstruct() {
+    this.registry.retry(SALADS_RETRY_NAME).getEventPublisher().onRetry(this::logRetry);
+    this.registry.retry(SALADS_RETRY_NAME).getEventPublisher().onSuccess(this::logSuccess);
+    this.registry.retry(SALADS_RETRY_NAME).getEventPublisher().onError(this::logError);
+  }
+
+  private void logRetry(final RetryOnRetryEvent event) {
+    log.log(INFO, "Bola by Kucapaca, ale skúsim zopakovať po {0}-krát!",
```
Comment (typekpb) [inline Retry/src/main/java/org/techsalad/retry/controller/SaladsController.java]: please use the english language only + log messages that are technically exact

File: Retry/src/main/java/org/techsalad/retry/controller/SaladsController.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        event.getNumberOfRetryAttempts());
+  }
+
+  private void logSuccess(final RetryOnSuccessEvent event) {
+    log.log(INFO, "Všetko dobre dopadlo po {0}-pokusoch :)", event.getNumberOfRetryAttempts());
```
Comment (typekpb) [inline Retry/src/main/java/org/techsalad/retry/controller/SaladsController.java]: dtto

File: Retry/src/main/java/org/techsalad/retry/controller/SaladsController.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  }
+
+  private void logError(final RetryOnErrorEvent event) {
+    log.log(WARNING, "Dačo hrozné! Urobilo bác po {0}-pokusoch! :(",
```
Comment (typekpb) [inline Retry/src/main/java/org/techsalad/retry/controller/SaladsController.java]: dtto

File: Retry/src/main/java/org/techsalad/retry/service/SaladsService.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+@Service
+public class SaladsService {
+
+  private int failuresCount = 0;
+
+  public List<String> getAllSalads() {
+    if (failuresCount < 3) {
+      failuresCount++;
+      throw new IllegalStateException("Bola kucapaca! Neviem najsť ani jeden šalát. :(");
```
Comment (typekpb) [inline Retry/src/main/java/org/techsalad/retry/service/SaladsService.java]: dto

File: Retry/src/main/java/org/techsalad/retry/service/SaladsService.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+    }
+    failuresCount = 0;
+
+    return Collections.singletonList("Najsamlepší šalátik!");
```
Comment (typekpb) [inline Retry/src/main/java/org/techsalad/retry/service/SaladsService.java]: dtto

File: Retry/src/test/java/org/techsalad/retry/controller/SaladsControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+  @Test
+  void testExecuteWithUnsuccessful5thAttempts() {
+    // given
+    when(this.saladsService.getAllSalads()).thenThrow(RUNTIME_EXCEPTION);
+
+    // when
+    final Executable getSalads = () -> saladsController.getSalads();
+
+    // then
+    assertThrows(RuntimeException.class, getSalads, "Hmm... tu to malo spadnut!??");
```
Comment (typekpb) [inline Retry/src/test/java/org/techsalad/retry/controller/SaladsControllerTest.java]: dtto

File: Retry/src/test/java/org/techsalad/retry/controller/SaladsControllerTest.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+@ExtendWith(SpringExtension.class)
+@SpringBootTest(
+    classes = {RetryAutoConfiguration.class, AnnotationAwareAspectJAutoProxyCreator.class,
+        SaladsController.class},
+    properties = {"resilience4j.retry.instances.getSalads.maxRetryAttempts=5",
+        "resilience4j.retry.instances.getSalads.retryExceptions=java.lang.RuntimeException",
+        "resilience4j.retry.instances.getSalads.waitDuration=1ms"})
+class SaladsControllerTest {
+
+  private static final RuntimeException RUNTIME_EXCEPTION = new RuntimeException("Zle, nedobre!");
```
Comment (typekpb) [inline Retry/src/test/java/org/techsalad/retry/controller/SaladsControllerTest.java]: dtto

File: Retry/src/main/java/org/techsalad/retry/service/SaladsService.java [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+
+import java.util.Collections;
+import java.util.List;
+
+import org.springframework.stereotype.Service;
+
+@Service
+public class SaladsService {
+
+  private int failuresCount = 0;
```
Comment (typekpb) [inline Retry/src/main/java/org/techsalad/retry/service/SaladsService.java:11]: as a bonus/optional step, we might want to refactor the sample in a way, that failing service would be part of unit test only.

as this means, we'll have to redeploy to demo again to simulate the failure.

potentially we should just call some external URL in the source code, that would be mocked for the test/demo purposes. Wiremock might be an option: https://wiremock.org/docs/stateful-behaviour/

**Review by typekpb — CHANGES_REQUESTED** [review]
any chance to use all lowercase? => `Retry` -> `retry`

File: Retry/src/main/java/org/techsalad/retry/service/SaladsService.java (hunk 9 above)
Comment (MVenglarcik) [reply Retry/src/main/java/org/techsalad/retry/service/SaladsService.java:11]: Ignored as jUnit tests covers also failure case + this is used for demonstrating the retry mechanism.

**Review by typekpb — APPROVED** [review]

**Pull request merged**